Add new fact for macOS Big Sur

poundbangbash opened the pull request:
BoardID and platform support gathered from the beta installer at 0dc2cd535db0da2a9f559215671686ea4c055394/AssetData/boot/PlatformSupport.plist

This may change as the beta rolls on, but getting this out there as a placeholder and assuming the hardware to test on will be the hardware that will support it later.

homebysix commented on facts/bigsur_upgrade_supported.py:
```diff
@@ -0,0 +1,251 @@
+
+
+def is_supported_system_version():
+    '''Returns True if current macOS version is 10.7 through 10.13,
```
Should say `through 10.13` should say `through 10.15`

homebysix commented on facts/bigsur_upgrade_supported.py:
```diff
@@ -0,0 +1,251 @@
+    macos_minor_version = get_minor_system_version()
+    if macos_minor_version >= 16:
+        return False
+    elif macos_minor_version >= 7:
```
According to the [Big Sur](https://apps.apple.com/us/app/macos-big-sur/id1526878132?mt=12) listing in the App Store, the minimum supported version should be 10.9, not 10.7.

homebysix commented on facts/bigsur_upgrade_supported.py:
```diff
@@ -0,0 +1,251 @@
+
+
+def fact():
+    '''Return our sierra_upgrade_supported fact'''
```
Cosmetic issue: Should say `bigsur_upgrade_supported` in this docstring (same typo occurs on previous versions of this fact too).

poundbangbash commented:
Thanks.  Updated the suggested edits.
Also, re-referenced supported models from Big Sur 11.0.1 (20B50)'s full installer (productID 001-83532, released Nov 19, 2020 at 1:02 PM) resources at /Volumes/Shared Support/com_apple_MobileAsset_MacSoftwareUpdate/com_apple_MobileAsset_MacSoftwareUpdate.xml. 

Removed support for MacBookPro11,1 modelIdentifier per https://support.apple.com/en-us/HT211242

flammable commented:
Thanks for this! I'm testing it in my org, but I'm running into some trouble. The condition is coming up `False` for an iMac that's showing the Big Sur upgrade in System Preferences.

I'm testing with an iMac (21.5-inch, Mid 2014). Here's the output when run directly:

```
is_virtual_machine:          False
get_current_model:           iMac14,4
is_supported_model:          False
get_minor_system_version:    15
is_supported_system_version: True
get_board_id:                Mac-81E3E92DD6088272
is_supported_board_id:       True
{'bigsur_upgrade_supported': False}
```

Should iMac14,4 be listed under `non_supported_models`?

poundbangbash commented:
Thanks for the feedback and testing, @flammable.  

The original munki-fact was crafted for macOS 10.12 and stated 
```
These checks are based on the installCheckScript and volCheckScript in distribution file of OSInstall.mpkg installer package. The OSInstall.mpkg can be found in the root of InstallESD disk image.
```

The `nonSupportedModels` variable existed up thru the macOS 10.15.7 installer.  However, the structure of the Big Sur installer is drastically different than previous OSen and doesn't provide specific _unsupported_ models, only supported `SupportedDeviceModels` which appear to be Bridge models and boardIDs.  I think for Big Sur we may need to move away from the `is_supported_model` info as that would have to be hand crafted instead of extracted from definitive data in the installer. The machine's `boardID` is specific information in the current Big Sur installers that can be compared to the client machine running this script and would result in the same end boolean being accurate as far as True/False.

I'll have to evaluate this and double back on this PR to make appropriate adjustments.

-Eric

poundbangbash commented:
Actually, digging around the installer payload I found a list of explicitly supported models and can use that.
```
Install macOS Big Sur.app/Contents/SharedSupport/SharedSupport.dmg - mount this
    /Volumes/Shared Support/com_apple_MobileAsset_MacSoftwareUpdate/da4c0b39d73549c809a57e9b9951e380b28b122d.zip - decompress this
        da4c0b39d73549c809a57e9b9951e380b28b122d/AssetData/boot/PlatformSupport.plist
```

The latest push to this PR incorporates those changes.  Try that code and let me know how it goes.

-Eric

poundbangbash commented:
Further testing locally shows that the list of supported models harvested from the PlatformSupport.plist isn't 100% accurate either.  My 16" MacBookPro16,1 isn't on the list but it is supported to install Big Sur.  

I've changed the evaluation of needing the supported model _AND_ supported boardID _AND_ the OS version needing to be `TRUE` to being a logical _OR_ for either the boardID and model and keeping the AND for OS version. Since if either the boardID or the model is supported, that should be enough to continue.

flammable commented:
Thanks, @poundbangbash! I just retested with both Macs and it's working great. 😄
